src/vscode-bicep/src/visualizer/view.ts

```diff
@@ -0,0 +1,211 @@
+  public render = debounce(() => this.doRender());
+
+  private async doRender() {
+    if (this.isDisposed || !this.readyForRender) {
+      return;
+    }
+
+    let document: vscode.TextDocument;
+    try {
+      document = await vscode.workspace.openTextDocument(this.documentUri);
+    } catch {
+      this.webviewPanel.webview.html = this.createDocumentNotFoundHtml();
+      return;
+    }
+
+    if (this.isDisposed) {
+      return;
+    }
+
+    const deploymentGraph = await this.languageClient.sendRequest(
+      deploymentGraphRequestType,
+      {
+        textDocument: this.languageClient.code2ProtocolConverter.asTextDocumentIdentifier(
+          document
+        ),
+      }
+    );
+
+    if (this.isDisposed) {
+      return;
+    }
+
+    this.webviewPanel.webview.postMessage(
+      createDeploymentGraphMessage(this.documentUri.fsPath, deploymentGraph)
+    );
+  }
+
+  private handleDidReceiveMessage(message: Message): void {
+    if (message.kind === "READY") {
+      this.readyForRender = true;
+      this.render();
+    }
+  }
+
+  private createWebviewHtml() {
+    const { cspSource } = this.webviewPanel.webview;
+    const nonce = BicepVisualizerView.getNonce();
+    const scriptUri = this.webviewPanel.webview.asWebviewUri(
+      vscode.Uri.joinPath(this.extensionUri, "out", "visualizer.js")
+    );
+
+    return `
+      <!DOCTYPE html>
+      <html lang="en">
+      <head>
+        <meta charset="UTF-8">
+        <!--
+        Use a content security policy to only allow loading images from https or from our extension directory,
+        and only allow scripts that have a specific nonce.
+        -->
+        <meta http-equiv="Content-Security-Policy" content="default-src 'none'; style-src ${cspSource} 'unsafe-inline'; img-src ${cspSource} https: data:; script-src 'nonce-${nonce}' vscode-webview-resource:;">
+        <meta name="viewport" content="width=device-width, initial-scale=1.0">
+        <title>Cat Scratch</title>
+      </head>
+      <body>
+        <div id="root"></div>
+        <script nonce="${nonce}">const vscode = acquireVsCodeApi();</script>
+        <script nonce="${nonce}" src="${scriptUri}" />
+
+      </body>
+      </html>`;
+  }
+
+  private createDocumentNotFoundHtml() {
+    const documentName = path.basename(this.documentUri.fsPath);
+
+    return `
+      <!DOCTYPE html>
+      <html lang="en">
+      <head>
+        <meta charset="UTF-8">
+        <meta name="viewport" content="width=device-width, initial-scale=1.0">
+      </head>
+      <body>
+        <div class="vscode-body">${documentName} not found. It might be deleted or renamed.</div>
+      </body>
+      </html>`;
+  }
+
+  private static getNonce() {
+    let text = "";
+    const possible =
+      "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
+    for (let i = 0; i < 32; i++) {
+      text += possible.charAt(Math.floor(Math.random() * possible.length));
+    }
+    return text;
+  }
```

Comment:
Unless there's some tangible security benefit from doing this (I'm assuming not since you're not using a crypto algorithm to generate the nonce), then would it be simpler to just hard-code the nonces for the two files?

Reply:
Good point on using a crypto algorithm! This was copied from the vscode sample repo without me thinking too much...I would prefer not to use hard-coded nonces though, because it will pretty much defeat the entire purpose of using the nonce. Changed this to use `crypto.randomBytes`.